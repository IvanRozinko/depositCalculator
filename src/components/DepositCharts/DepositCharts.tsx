import React from 'react'
import { Box, Divider, Button } from '@material-ui/core'; 
 import { Theme, makeStyles, createStyles } from '@material-ui/core/styles';
import { useDepositContext } from '../../context/DepositInputs';
import InfoRow from '../InfoRow/InfoRow';
import IconTooltip from '../IconTooltip/IconTooltip';
import ProfitDetails from '../ProfitDetails/ProfitDetails';
import Charts from '../Charts/Charts';

import { uahFormatter } from '../../helpers/helpers';
import { DEPOSIT, DEPOSIT_TAX } from '../../config/constants';
import Send from '@material-ui/icons/SendOutlined';

function DepositCharts() {
  const useStyles = makeStyles((theme: Theme) => 
    createStyles({
      container: {
        paddingLeft: 20,
        [theme.breakpoints.down('sm')]: {
          paddingLeft: 0,
          paddingTop: 20
        }
      },
      mb: {
        marginBottom: 20
      },
      buttonContainer: {
        [theme.breakpoints.down('sm')]: {
          display: 'flex',
          justifyContent: 'center'
        }
      }
    })
  );
  const styles = useStyles();
  const { deposit } = useDepositContext();

  const { 
    monthly, withoutFirstMonth, duration, percent, sum, percentDestination
   } = deposit;

  const monthlyTotal = () => withoutFirstMonth 
    ? monthly * ( duration - 1 ) 
    : monthly * duration
  
  const roundTwoSigns = (number: number) => Math.round(number * 100) / 100;

  const sumProfit = new Array(duration)
    .fill(0)
    .reduce((acc) => { 
      const montlyProfit = roundTwoSigns(acc.sum * percent / 100 / 12);
      const raisedSum = roundTwoSigns(acc.sum + monthly);
      return {
        sum: percentDestination === DEPOSIT ? raisedSum + montlyProfit : raisedSum,
        profit: montlyProfit + acc.profit
      }
    } 
    , { 
      sum: withoutFirstMonth ? sum : sum + monthly, // counting raised sum starting first month
      profit: 0 
    }
  );

  const handleStartDeposit = () => alert('Є зайві гроші?');

  const tax = sumProfit.profit * DEPOSIT_TAX;
  const purePercent = percent - percent * DEPOSIT_TAX;
  const pureProfit = uahFormatter(sumProfit.profit - tax);
  return (
    <Box 
      display='flex' 
      flex={1}
      flexDirection='column'
      className={styles.container}
    >
      <Charts 
        sum={sum}
        monthlyTotal={monthlyTotal()}
        profit={sumProfit.profit}
      />
      <Box>
        <InfoRow label='Сума вкладу' value={uahFormatter(deposit.sum)} />
        <InfoRow label='Сума поповнень' value={uahFormatter(monthlyTotal())} />
        <InfoRow label='Процентна ставка' value={`${percent} % на рік`} />
        <InfoRow label='Проценти (дохід)' value={uahFormatter(sumProfit.profit)} valueBold />
        <InfoRow label='Податок' value={uahFormatter(tax)} children={<IconTooltip />} />
        <ProfitDetails purePercent={purePercent} pureProfit={pureProfit} />
        <Divider className={styles.mb} />
        <Box className={styles.buttonContainer}>
          <Button
            variant='contained'
            color='primary'
            startIcon={<Send />}
            onClick={handleStartDeposit}
          >
            Оформити депозит
          </Button>
        </Box>
      </Box>
    </Box>
  )
}

export default DepositCharts
