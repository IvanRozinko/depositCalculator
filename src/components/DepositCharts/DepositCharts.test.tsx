import DepositCharts from "./DepositCharts";
import { shallow } from 'enzyme';
import { Button } from '@material-ui/core';

describe('<DepositChats />', () => {
  const wrapper = shallow(
    <DepositCharts />
  )

  it('should renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  })

  it('should show alert on start Deposit', () => {
    global.alert = jest.fn();
    wrapper.find(Button).simulate('click');
    expect(window.alert).toBeCalled();
  })
})