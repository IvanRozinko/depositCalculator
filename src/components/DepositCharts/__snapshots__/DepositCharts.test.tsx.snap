// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DepositChats /> should renders correctly 1`] = `
<Styled(MuiBox)
  className="makeStyles-container-1"
  display="flex"
  flex={1}
  flexDirection="column"
>
  <Charts
    monthlyTotal={40000}
    profit={2517.55}
    sum={100000}
  />
  <Styled(MuiBox)>
    <InfoRow
      label="Сума вкладу"
      value="100 000,00 ₴"
    />
    <InfoRow
      label="Сума поповнень"
      value="40 000,00 ₴"
    />
    <InfoRow
      label="Процентна ставка"
      value="6 % на рік"
    />
    <InfoRow
      label="Проценти (дохід)"
      value="2 517,55 ₴"
      valueBold={true}
    />
    <InfoRow
      label="Податок"
      value="490,92 ₴"
    >
      <IconTooltip />
    </InfoRow>
    <ProfitDetails
      purePercent={4.83}
      pureProfit="2 026,63 ₴"
    />
    <WithStyles(ForwardRef(Divider))
      className="makeStyles-mb-2"
    />
    <Styled(MuiBox)
      className="makeStyles-buttonContainer-3"
    >
      <WithStyles(ForwardRef(Button))
        color="primary"
        onClick={[Function]}
        startIcon={<Memo />}
        variant="contained"
      >
        Оформити депозит
      </WithStyles(ForwardRef(Button))>
    </Styled(MuiBox)>
  </Styled(MuiBox)>
</Styled(MuiBox)>
`;
